### ISSUE TITLE
MaxInFlight filter should not limit controlling requests like leader election

### ISSUE THREAD

**Issue by kevin-wangzefeng** [kubernetes/kubernetes#49472]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

Currently maxinflight filter only skip tracking long running requests, controlling requests like leader election are tracked and limited together with other data requests.
Leader election would therefore fail when there are too many requests trying to fetch some data.

**What you expected to happen**:

Skip limiting leader election and other controlling requests (if we can clarify) , or, track/limit them separately.

**Comment by liggitt** [comment]
> Skip limiting leader election and other controlling requests (if we can clarify) , or, track/limit them separately.

Leader election has been done on top of regular endpoints/configmap resources. The apiserver should not special case specific names of resources and exempt them from normal filter chain handling.

**Comment by liggitt** [comment]
Rather than exempt specific resources, it would make sense to bucket rate limiting by the requesting user (or class of requesting user), with control plane components given higher limits

**Comment by sttts** [comment]
Wasn't there a PR some months ago about throttling per user (= ip address) ?

Moreover, more fine grained rate limiting is a topic in big clusters. We had issues with LIST requests in those situation saturating the apiserver. The only way to keep it from panicing due to request timeouts (60sec) was to reduce MaxInFlight, influencing all other not to costly, but time critical operation as well, compare https://bugzilla.redhat.com/show_bug.cgi?id=1439324.

In other words, rather than ad-hoc solutions for single problems like leader election I think our throttling needs deeper thought and design.

**Comment by kevin-wangzefeng** [comment]
@liggitt @sttts Thanks for the comments, I agree that we should not make ad-hoc solutions for single problems like leader election. BTW, is there any existing discussion threads on more fine grained rate limiting in big clusters?

**Comment by caesarxuchao** [comment]
cc @jpbetz @cheftako

**Comment by dims** [comment]
long-term-issue (note to self)
